compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -1785,3 +1796,36 @@ pub fn is_type_without_fields(item: &Annotatable) -> bool {
         false
     }
 }
+
+/// A folder used to correctly mark every span coming from a `derive` macro.
+struct SpanMarker {
+    container_id: rustc_span::hygiene::LocalExpnId,
+    span: Span,
+}
+
+    // let mut marker = Marker(cx.current_expansion.id, transparency);
+impl MutVisitor for SpanMarker {
+    fn visit_span(&mut self, sp: &mut Span) {
+        // let x = sp.ctxt();
+        // let y = self.container_id.to_expn_id();
+        // tracing::info!(?sp, ?x, ?self.container_id, ?y);
+        let x = self.span.ctxt().outer_expn();
+        let y = sp.ctxt().outer_expn();
+        tracing::info!(?self.span, ?sp, ?self.container_id, ?x, ?y);
+        // if self.span != *sp {
+            // *sp = sp.apply_mark(self.container_id.to_expn_id(), rustc_span::hygiene::Transparency::SemiTransparent);
+        //     // *sp = sp.with_ctxt(self.span.ctxt());
+        // } else {
+            // *sp = sp.with_ctxt(self.span.ctxt());
+        // }
+        if self.container_id.to_expn_id() != sp.ctxt().outer_expn() { // && sp.ctxt().outer_expn() == rustc_span::hygiene::ExpnId::root() {
+            // println!("{:?} {:?} {:?}", self.ctxt, sp.ctxt(), sp);
+            // tracing::info!("marked");
+            // *sp.with_ctxt();
+            *sp = sp.apply_mark(self.container_id.to_expn_id(), rustc_span::hygiene::Transparency::SemiTransparent);
```

Comment:
This is not what I was talking about in #90519, all the marks are already added in the `TraitDef::span` span (`self.span` in `fn create_derived_impl`).

This visitor only needs to replace the visited span's context with the `TraitDef::span` context while keeping the location.
```rust
fn visit_span(&mut self, span: &mut Span) {
    *span = span.with_ctxt(self.self_span_from_TraitDef);
}
```

Reply:
Ah, of course, some parts of the output should not be visited, after all.

The `Foo` parts, specifically, i.e. references to the type for which we are deriving the impl.
```rust
#[derive(Default)]
struct Foo;
```
```rust
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::core::default::Default for Foo {
    #[inline]
    fn default() -> Foo { Foo{} }
}
```